rfc-160-using-es6-modules-in-govuk_publishing_components.md

```diff
@@ -0,0 +1,428 @@
+
+If we were to move the Javascript to be transpiled to `app/assets/javascripts/govuk_publishing_components/components/src/` and build the output in `app/assets/javascripts/govuk_publishing_components/components/` (as shown in the config above) it would mean we could make this change to govuk_publishing_components without having to update paths in frontend rendering applications. This is a [similar folder structure to what is used in govuk-frontend](https://github.com/alphagov/govuk-frontend/tree/main/packages/govuk-frontend/src). If analytics were also to be built by the bundler then a similar change to folder structure should be made (and those paths could be added to a [globStream](https://github.com/isaacs/node-glob#globstreampattern-string--string-options-globoptions--minipassstring--path)) although I would argue that [analytics does not necessarily need to be updated](#loading-analytics-code-individually). This script would be run on gem release so that an application could then include the generated Javascript files (similar to how node_modules are included).
+
+##### 4. Update frontend applications to individually load Javascript
+
+In order to load the JS for each component separately, we need to add the path of each component we are going to use to the manifest of a frontend application. Then we can add a script tag to the pages were the components are going to be used:
+
+```
+//= link govuk_publishing_components/path/to/component-name
+```
+
+then
+
+```
+  <%= javascript_include_tag "govuk_publishing_components/path/to/component-name" %>
+```
+
+If we were to follow this implementation method then we could augment AssetHelper instead of manually adding script tags to partials and paths to the manifest. AssetHelper already has this functionality for stylesheets, where the `add_component_stylesheet` function can update the manifest and add stylesheet link tags. We could add a new function for adding javascript to a partial:
+
+`add_component_javascript("component-name")`
+
+which would add
+
+`<script type="module" src="path/to/component"/>`
+
+similar to the format of `add_component_stylesheet("component-name")`. As the latest version of govuk-frontend is targeting browsers that support `type="module"` for components we use a script tag with the type module. Although analytics could also be individually loaded, [I would argue that for the intermediary change this is not required](#loading-analytics-code-individually).
+
+[As well as the ability to test the support of a script, there are also performance benefits to individual loading of Javascript. In the performance tests in the appendix, I examine the differences in page load time and asset size if we were to switch to individually loaded Javascript](#performance-tests-of-Individual-javascript-loading).
+
+### Significant change: Move all Javascript to a separate npm module
```

Comment:
I'm concerned about separating the JS out into a separate NPM package. You've highlighted some of the problems with this already (maintaining a separate package, how would local development work) but I think it's worth clarifying further.

Local development for frontend devs is already pretty complicated at times. For example (and I need to do this frequently) running a local application pointed at a local static, both of them pointing at a local version of `govuk_publishing_components`. This can be further complicated by any of those local apps randomly caching asset files, which leads to a lot of time wasted hunting through to delete them, or rebuild or restart applications. Adding another layer into this (a local version of all the JS for `govuk_publishing_components`) is going to make things more complicated.

It's also worth thinking about the process for making and deploying a change.

- make change to a component in `govuk_publishing_components` (1 PR)
- make change to component JS in JS package (+1 PR)
- make change in application and static to use component change (+2 PRs)
- make new version of JS package and new version of `govuk_publishing_components` to include it (+2 PRs)
- install new version of `govuk_publishing_components` in static and applications (+2 PRs, possibly more)

We're looking at potentially 8 pull requests for a single change, and even more if that change turns out to have an unforeseen bug.

We also have a problem (currently unsolved) in `govuk_publishing_components` of JS test files having to contain their own copy of the HTML that they operate on. If the JS and the component template are modified but the HTML in the test is not, it is possible to introduce errors into the code but still have the tests pass (again this has happened several times). Separating the JS and JS tests from the components they operate on still further by moving them out of `govuk_publishing_components` removes any chance of solving this problem.

I think all of this might come down to the question of whether separating out the JS into a separate package is something we _should_ do, or something we _have_ to do. 

If we have to do it then we have to, but this seems to be based on an assumption about the future of Rails that hasn't been confirmed yet. I'd also be interested to investigate if there are other approaches that could work for us even if that's what Rails decides.

As you mentioned, this issue might be worth a separate RFC.

Reply:
Can you keep the npm package in the same repo?

We use [npm workspaces](https://docs.npmjs.com/cli/v9/using-npm/workspaces) on GOV.UK Frontend with all our tools and [review app](https://govuk-frontend-review.herokuapp.com) as separate packages:

**package.json**
```json
"workspaces": [
  ".github/workflows/scripts",
  "packages/govuk-frontend",
  "packages/govuk-frontend-review",
  "shared/*",
  "docs/examples/*"
],
```

1. [**Publishable** packages in `./packages`](https://github.com/alphagov/govuk-frontend/tree/main/packages)
2. [**Internal-only** packages in `./shared`](https://github.com/alphagov/govuk-frontend/tree/main/shared)

Makes sure things like `import 'govuk-frontend'` hit the local package, not the registry, so no publishing needed. You'll notice when running `npm install` they'll be symbolically linked into `node_modules` too

But you get other [npm workspaces](https://docs.npmjs.com/cli/v9/using-npm/workspaces) features like:

### Build all workspaces
```console
npm run build --workspaces
```

### Build single workspce
```console
npm run build --workspace govuk-frontend
```

Reply:
Just catching up on this RFC (super late) and thought I should add a comment here as I imagine this suggestion was a misunderstanding based on a suggestion I made.

The idea of using a separate repo was the misunderstanding, that should not have been a considered option. It should have been exactly what @colinrotherham described where the same repo publishes both a gem and an npm module. This is what Rails core gems do (for example: https://github.com/rails/rails/tree/main/activestorage).

Not that doing this is without headaches, but vastly less headaches than an idea of two separate codebases.